integrations/google-analytics-4/lib/index.js

```diff
@@ -0,0 +1,275 @@
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+var integration = require('@segment/analytics.js-integration');
+var reject = require('reject');
+
+/**
+ * GA4
+ */
+var GA4 = (module.exports = integration('Google Analytics 4')
+  .global('gtag')
+  .global('ga4DataLayer')
+  .option('measurementIds', [])
+  .option('cookieDomainName', 'auto')
+  .option('cookiePrefix', '_ga')
+  .option('cookieExpiration', 63072000)
+  .option('cookieUpdate', true)
+  .option('cookieFlags', '')
+  .option('disablePageViewMeasurement', true)
+  .option('disableAllAdvertisingFeatures', false)
+  .option('disableAdvertisingPersonalization', false)
+  .option('disableGoogleAnalytics', false)
+  .option('sendUserId', false)
+  .option('userProperties', {})
+  .option('customEventsAndParameters', [])
+  .tag(
+    '<script src="//www.googletagmanager.com/gtag/js?id={{ measurementId }}&l=ga4DataLayer">'
+  ));
+
+/**
+ * Initialize.
+ *
+ * https://developers.google.com/analytics/devguides/collection/ga4
+ *
+ * @api public
+ */
+GA4.prototype.initialize = function() {
+  window.ga4DataLayer = window.ga4DataLayer || [];
+  window.gtag = function() {
+    window.ga4DataLayer.push(arguments);
+  };
+
+  /**
+   * This line is in all of the gtag examples but is not well documented. Research
+   * says that it is is related to deduplication.
+   * https://stackoverflow.com/questions/59256532/what-is-the-js-gtags-js-command
+   */
+  window.gtag('js', new Date());
+
+  var opts = this.options;
+  var measurementIds = opts.measurementIds;
+
+  /**
+   * Avoid loading and configuring gtag.js if any are true:
+   *  - Disable Google Analytics setting is enabled
+   *  - No measurement IDs are configured
+   */
+  if (!measurementIds.length || opts.disableGoogleAnalytics) {
+    return;
+  }
+
+  var config = {
+    /**
+     * Disable Automatic Page View Measurement
+     * https://developers.google.com/analytics/devguides/collection/ga4/disable-page-view
+     */
+    send_page_view: !opts.disablePageViewMeasurement,
+
+    /**
+     * Cookie Update
+     * https://developers.google.com/analytics/devguides/collection/ga4/cookies-user-id#cookie_update_parameter
+     */
+    cookie_update: opts.cookieUpdate,
+
+    /**
+     * Cookie Domain Name
+     * https://developers.google.com/analytics/devguides/collection/ga4/cookies-user-id#cookie_domain_configuration
+     */
+    cookie_domain: opts.cookieDomainName,
+
+    /**
+     * Cookie Prefix
+     * https://developers.google.com/analytics/devguides/collection/ga4/cookies-user-id#cookie_prefix
+     */
+    cookie_prefix: opts.cookiePrefix,
+
+    /**
+     * Cookie Expiration
+     * https://developers.google.com/analytics/devguides/collection/ga4/cookies-user-id#cookie_expiration
+     */
+    cookie_expires: opts.cookieExpiration,
+  };
+
+  var sets = [
```

Comment:
Should this be an array of objects? A little bit confused by the different format of Cookie Flags and the rest of them. Could be missing something

Reply:
gtag.js has a [`set` command](https://developers.google.com/gtagjs/reference/api#set) and it's really confusing. Sometimes it accepts a [single object argument](https://developers.google.com/analytics/devguides/collection/ga4/cookies-user-id#cookie_flags) (L1) and sometimes it accepts [multiple arguments](https://developers.google.com/analytics/devguides/collection/ga4/display-features#disable_all_advertising_features) (L2):
```
window.gtag('set', { cookie_flags: opts.cookieFlags })
window.gtag('set', 'allow_google_signals', !opts.disableAllAdvertisingFeatures)
```

The `sets` var we have is an array of arrays to represent a list of set commands and the arguments to pass.

Reply:
Gotcha makes sense. Thanks for clarifying